src/uu/rm/src/rm.rs

```diff
@@ -442,46 +468,137 @@ fn remove_file(path: &Path, options: &Options) -> bool {
+                                &(format!(
+                                    "remove write-protected regular file {}? ",
+                                    path.quote()
+                                )),
+                            )
+                        }
+                    } else if options.interactive == InteractiveMode::Always {
+                        if metadata.len() == 0 {
+                            prompt(&(format!("remove regular empty file {}? ", path.quote())))
+                        } else {
+                            prompt(&(format!("remove file {}? ", path.quote())))
+                        }
+                    } else {
+                        true
+                    }
                 } else {
-                    if fs::metadata(path).unwrap().len() == 0 {
-                        return prompt(
-                            &(format!(
-                                "rm: remove write-protected regular empty file {}? ",
-                                path.quote()
-                            )),
-                        );
+                    true
+                }
+            }
+            Err(err) => {
+                if err.kind() == ErrorKind::PermissionDenied {
+                    if let Ok(metadata) = fs::metadata(path) {
+                        if metadata.len() == 0 {
+                            prompt(
+                                &(format!(
+                                    "remove write-protected regular empty file {}? ",
+                                    path.quote()
+                                )),
+                            )
+                        } else {
+                            prompt(
+                                &(format!(
+                                    "remove write-protected regular file {}? ",
+                                    path.quote()
+                                )),
+                            )
+                        }
+                    } else {
+                        prompt(&(format!("remove write-protected regular file {}? ", path.quote())))
                     }
-                    prompt(&(format!("rm: remove write-protected regular file {}? ", path.quote())))
+                } else {
+                    true
                 }
-            } else {
-                true
             }
         }
     }
 }
 
-fn prompt_descend(path: &Path) -> bool {
-    prompt(&(format!("rm: descend into directory {}? ", path.quote())))
+// For directories finding if they are writable or not is a hassle. In Unix we can use the built-in rust crate to to check mode bits. But other os don't have something similar afaik
+#[cfg(unix)]
+fn handle_writable_directory(path: &Path, options: &Options, metadata: &Metadata) -> bool {
+    use std::os::unix::fs::PermissionsExt;
+    let mode = metadata.permissions().mode();
+    // Check if directory has user write permissions
+    // Why is S_IWUSR showing up as a u16 on macos?
```

Comment:
This is expected, `libc` defines the modes as the type `mode_t`, which is `u32` on linux and `u16` on mac. You can check it here: https://docs.rs/libc/latest/libc/type.mode_t.html (and by changing the platform in the top-left corner)

Reply:
The permission check seems different from before now? I think we need something like this:

https://github.com/uutils/coreutils/blob/2021c56dd0db4d6dad2fe03d0d57b5e5595fd168/src/uu/test/src/test.rs#L366-L374

Could that be correct? I have to admit I'm not an expert on this, so I could be wrong :)

Reply:
Looking at there's it looks to be the same as what I was doing before with shifting of bits with that magic number

Reply:
Since for rm we don't care about read or execute permission we don't need to use a Permission enum like they use and because I found the libc constant I don't need to use the bit shifting... I believe

Reply:
I don't think it's the same. It feels like we should do something like what test is doing (though indeed only for writing). `S_IWUSR` is documented as "write permissions for the owner of the file" so that does not say much about the current user or access by group. Several sources also recommend using `libc::access` (or `nix::access`). @sylvestre do you know how this all works and what the right approach is?

Reply:
The default rm only checks for owner write permission you can test that by running:
```
touch u-w-empty.txt
touch g-w-empty.txt
touch o-w-empty.txt
chmod u-w u-w-empty.txt
chmod g-w g-w-empty.txt
chmod o-w o-w-empty.txt
```
Then use `rm *`
rm then only prompts for u-w-empty.txt
`rm: remove write-protected regular empty file 'u-w-empty.txt'? y`

Reply:
Hmmm, you seem to be right. I think you've got a nice approximation at the moment. Looking at the GNU code this whole thing is quite complicated. Let's put a comment here that it's probably not 100% compatible and move on with our lives 😄